tests/shared/src/test/scala/cats/tests/SyntaxSuite.scala

```diff
@@ -246,9 +246,18 @@ object SyntaxSuite {
     val fb = mock[M[B]]
     val fc = mock[M[C]]
     val f = mock[(A, B, C) => Z]
+    val mf = mock[(A, B, C) => M[Z]]
 
     tfabc.parMapN(f)
     (fa, fb, fc).parMapN(f)
+
+    tfabc.parFlatMapN(mf)
+    (fa, fb, fc).parFlatMapN(mf)
+
+    val tfa = mock[Tuple1[M[A]]]
+    val mfone = mock[A => M[Z]]
+
```

Comment:
We aren't exercising parFlatMap2 etc are we?

Should we at least exercise a few of the specific arity functions?

Reply:
Let me understand: you mean testing i.e. `Parallel.parFlatMap2(fa, fb)(mftwo)` in the test? (or `P.parFlatMap2(fa, fb)(mftwo)`)

To the best of my understanding the `parMap$arity` and `parFlatMap$arity` functions are created just in the `ParallelArityFunctions` trait that is just mixed in `Parallel`, while in the single syntax classes for each arity of tuple the exposed methods are `mapN`, `parMapN`, `flatMapN` and now `parFlatMapN`, so the syntax (i.e.) `(fa, FB).parFlatMap2(mftwo)` is not available.

OTOH, this will be the only test in the syntax suite that tests this kind of syntax, am I right?

Reply:
I would say that tests for `parFlatMap2`, `parFlatMap3`, etc do not belong to `SyntaxSuite` because it is not about testing some sort of extension syntax.

That said, I think it may make sense to add a couple of tests to `ParallelSuite` to make sure that `parFlatMap2` is working as expected.

Reply:
Okay, that makes sense :D

Reply:
@johnynek I added some tests to prove consistency between parFlatMapN and mapN.flatten, let me know if it works for you.

Reply:
My concern is that `parFlatMap2, parFlatMap3, ...` are totally unexercised by our tests. There is a difference with parMapN since parMapN actually delegates to `parMap2, parMap3, ...` so that code is called.

You went a different route with this code choosing instead to inline a different definition.

So, I think we should either call the same code so we do have test coverage, or explicitly test at least 2 of the generated functions.

Reply:
Put another way: I think we should virtually never add untested code to this repo. I think our goal should be 100% test coverage as necessary, but not sufficient, for confidence that we are minimizing our bug count.

Reply:
That's a good point @johnynek, and now that you say it I think I prefer to change the implementation of `parFlatMapN` from `p.flatMap.flatten(Parallel.parMap$arity($tupleArgs)(f))` to `Parallel.parFlatMap$arity($tupleArgs)(f)"`, what do you think? In this way, we can avoid adding specific tests for (theoretically) every `Parallel` instance plus the heavy lifting of the method's logic is not done in the parFlatMapN implementation but rather in every parFlatMap$arity.

Reply:
Something like this 👇 . I already tested it locally and it works indeed.
```diff
diff --git a/project/Boilerplate.scala b/project/Boilerplate.scala
index 111f6f542..50f03432c 100644
--- a/project/Boilerplate.scala
+++ b/project/Boilerplate.scala
@@ -479,8 +479,7 @@ object Boilerplate {
         if (arity == 1)
           s"def parFlatMap[Z](f: (${`A..N`}) => M[Z])(implicit p: NonEmptyParallel[M]): M[Z] = p.flatMap.flatMap($tupleArgs)(f)"
         else
-          s"def parFlatMapN[Z](f: (${`A..N`}) => M[Z])(implicit p: NonEmptyParallel[M]): M[Z] = p.flatMap.flatten(Parallel.parMap$arity($tupleArgs)(f))"
-
+          s"def parFlatMapN[Z](f: (${`A..N`}) => M[Z])(implicit p: NonEmptyParallel[M]): M[Z] = Parallel.parFlatMap$arity($tupleArgs)(f)"
       val parTupled =
         if (arity == 1) ""
         else
```

Reply:
@johnynek I pushed the implementation that I documented here ☝️. As you suggested, this might test a couple of arity instances since the parFlatMapN implementation now relies on those. May I ask you and other reviewers like @armanbilge, @danicheg and @satorg to give me feedback, please?